src/main/java/com/mojang/serialization/MapCodec.java

```diff
@@ -0,0 +1,290 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+package com.mojang.serialization;
+
+import com.mojang.datafixers.DataFixUtils;
+import com.mojang.datafixers.util.Pair;
+import com.mojang.serialization.codecs.RecordCodecBuilder;
+
+import java.util.function.BiFunction;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.Stream;
+
+public abstract class MapCodec<A> extends MapDecoder.Implementation<A> implements MapEncoder<A> {
```

Comment:
`MapDecoder.Implementation` and `MapEncoder.implementation` have same contents. I would suggest pulling that to base class (so this just becomes `MapCodec<A> extends MapCompressorProvider<A> implements MapDecoder<A>, MapEncoder<A>`)

Reply:
:+1: